src/diagram/split_serialized.rs

```diff
@@ -0,0 +1,583 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+use std::{any::TypeId, collections::HashMap, usize};
+
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tracing::debug;
+
+use crate::{
+    Builder, Chain, ForRemaining, FromSequential, FromSpecific, ListSplitKey, MapSplitKey,
+    OperationResult, SplitDispatcher, Splittable,
+};
+
+use super::{
+    impls::{DefaultImpl, NotSupported},
+    join::register_join_impl,
+    register_serialize, DiagramError, DynOutput, NextOperation, NodeRegistry, SerializeMessage,
+};
+
+#[derive(Debug, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct SplitOp {
+    #[serde(flatten)]
+    pub(super) params: SplitOpParams,
+
+    pub(super) remaining: Option<NextOperation>,
+}
+
+#[derive(Debug, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum SplitOpParams {
+    Index(Vec<NextOperation>),
+    Key(HashMap<String, NextOperation>),
+}
+
+impl Splittable for Value {
+    type Key = MapSplitKey<String>;
+    type Item = (JsonPosition, Value);
+
+    fn validate(_: &Self::Key) -> bool {
+        true
+    }
+
+    fn next(key: &Option<Self::Key>) -> Option<Self::Key> {
+        MapSplitKey::next(key)
+    }
+
+    fn split(self, mut dispatcher: SplitDispatcher<'_, Self::Key, Self::Item>) -> OperationResult {
+        match self {
+            Value::Array(array) => {
+                for (index, value) in array.into_iter().enumerate() {
+                    let position = JsonPosition::ArrayElement(index);
+                    match dispatcher.outputs_for(&MapSplitKey::Sequential(index)) {
+                        Some(outputs) => {
+                            outputs.push((position, value));
+                        }
+                        None => {
+                            if let Some(outputs) = dispatcher.outputs_for(&MapSplitKey::Remaining) {
+                                outputs.push((position, value));
+                            }
+                        }
+                    }
+                }
+            }
+            Value::Object(map) => {
+                let mut next_seq = 0;
+                for (name, value) in map.into_iter() {
+                    let key = MapSplitKey::Specific(name);
+                    match dispatcher.outputs_for(&key) {
+                        Some(outputs) => {
+                            let position = JsonPosition::ObjectField(key.specific().unwrap());
```

Comment:
I know I wrote this line, but I had already forgotten what the safety rationale is for being able to use `.unwrap()` here. We should add a line of documentation like

```rust
// SAFETY: This key was initialized as MapSplitKey::Specific earlier in the function and is immutable,
// so this method is guaranteed to return Some
```